pkg/registry/core/service/storage/storage.go

```diff
@@ -648,9 +654,19 @@ func needsClusterIP(svc *api.Service) bool {
 }
 
 func needsNodePort(svc *api.Service) bool {
-	if svc.Spec.Type == api.ServiceTypeNodePort || svc.Spec.Type == api.ServiceTypeLoadBalancer {
+	if svc.Spec.Type == api.ServiceTypeNodePort {
 		return true
 	}
+	if svc.Spec.Type == api.ServiceTypeLoadBalancer {
+		if svc.Spec.AllocateLoadBalancerNodePorts == nil || *svc.Spec.AllocateLoadBalancerNodePorts {
+			return true
+		}
```

Comment:
at this point it was already defaulted IIRC https://github.com/kubernetes/kubernetes/blob/6aac45ff1e99068e834ba3b93b673530cf62c007/pkg/apis/core/v1/defaults.go#L139-L143

Reply:
we need the nil check because the field was imported later and the legacy objects don't have the filed. and the unmarshal gives it a `null` and bring apiserver nil pointer panics.